File: cpp/cpp-std/exer07c-data-server.cpp
/*
THE DATA SERVER PROBLEM
Version C: Solving the problem using a count-down latch
*/


#include <iostream>
#include <string>
#include <vector>
#include <chrono>
#include <thread>
#include <latch>
using namespace std;



#define sleepsec(secs) \
    do { std::this_thread::sleep_for(std::chrono::seconds(secs)); } while (0)



void checkAuthUser() {
    cout << "[   Auth   ] Start" << endl;
    // Send request to authenticator, check permissions, encrypt, decrypt...
    sleepsec(20);
    cout << "[   Auth   ] Done" << endl;
}



void processFiles(const vector<string>& lstFileName, std::latch& rdLatch) {
    for (auto&& fileName : lstFileName) {
        // Read file
        cout << "[ ReadFile ] Start " << fileName << endl;
        sleepsec(10);
        cout << "[ ReadFile ] Done  " << fileName << endl;

        rdLatch.count_down();

        // Write log into disk
        sleepsec(5);
        cout << "[ WriteLog ]" << endl;
    }
}



void processRequest() {
    const vector<string> lstFileName = { "foo.html", "bar.json" };
    std::latch readFileLatch(lstFileName.size());

    // The server checks auth user while reading files, concurrently
    std::thread th(&processFiles, std::cref(lstFileName), std::ref(readFileLatch));
    checkAuthUser();

    // The server waits for completion of loading files
    readFileLatch.wait();

    cout << "\nNow user is authorized and files are loaded" << endl;
    cout << "Do other tasks...\n" << endl;

    th.join();
}



int main() {
    processRequest();
    return 0;
}

File: csharp/exercise/exer07c-data-server.cs
/*
 * THE DATA SERVER PROBLEM
 * Version C: Solving the problem using a count-down latch
 */
using System;
using System.Threading;



class Exer07C : IRunnable
{
    public void run()
    {
        var server = new DataServer();
        server.processRequest();
    }


    private class DataServer
    {
        public void processRequest()
        {
            var lstFileName = new String[] { "foo.html", "bar.json" };
            var readFileLatch = new CountdownEvent(lstFileName.Length);

            // The server checks auth user while reading files, concurrently
            new Thread(() => processFiles(lstFileName, readFileLatch)).Start();
            checkAuthUser();

            // The server waits for completion of loading files
            readFileLatch.Wait();

            Console.WriteLine("\nNow user is authorized and files are loaded");
            Console.WriteLine("Do other tasks...\n");
        }


        // This task consumes CPU (and network bandwidth, maybe)
        private void checkAuthUser()
        {
            Console.WriteLine("[   Auth   ] Start");
            // Send request to authenticator, check permissions, encrypt, decrypt...
            Thread.Sleep(20000);
            Console.WriteLine("[   Auth   ] Done");
        }


        // This task consumes disk
        private void processFiles(String[] lstFileName, CountdownEvent latch)
        {
            foreach (var fileName in lstFileName)
            {
                // Read file
                Console.WriteLine("[ ReadFile ] Start " + fileName);
                Thread.Sleep(10000);
                Console.WriteLine("[ ReadFile ] Done  " + fileName);

                latch.Signal();

                // Write log into disk
                Thread.Sleep(5000);
                Console.WriteLine("[ WriteLog ]");
            }
        }
    }
}
